namedisl/__init__.py

```diff
@@ -32,72 +35,612 @@
+
+    if isinstance(obj._obj, isl.MultiAff | isl.PwMultiAff):
+        raise ValueError(
+        "Must align each component of a MultiExpression individually")
+
+    new_isl_obj = obj._obj.copy()
+    new_name_to_dim = dict(obj._name_to_dim)
+
+    for name, (dim_type, pos) in sorted(ordering.items(), key=lambda k: k[1][1]):
+
+        if name in obj._name_to_dim:
+            _, old_pos = new_name_to_dim[name]
+            if old_pos == pos:
+                    continue
+
+            temp_dim_type = isl.dim_type.param
+            if temp_dim_type == dim_type:
+                temp_dim_type = isl.dim_type.set
+
+            temp_pos = new_isl_obj.dim(temp_dim_type)
+
+            new_isl_obj = new_isl_obj.move_dims(
+                temp_dim_type, temp_pos, dim_type, old_pos, 1)
+            new_isl_obj = new_isl_obj.move_dims(
+                dim_type, pos, temp_dim_type, temp_pos, 1)
+        else:
+            old_pos = new_isl_obj.dim(dim_type)
+            new_isl_obj = new_isl_obj.insert_dims(dim_type, pos, 1)
+
+            if dim_type == isl.dim_type.param:
+                new_isl_obj = new_isl_obj.set_dim_name(dim_type, pos, name)
+
+        temp_name_to_dim = new_name_to_dim.copy()
+        for cur_name, (cur_dt, cur_pos) in sorted(new_name_to_dim.items(),
+                                                  key=lambda k: k[1][1]):
+            if cur_dt != dim_type:
+                continue
+
+            if (pos > old_pos) and (cur_pos > old_pos):
+                temp_name_to_dim[cur_name] = (cur_dt, cur_pos - 1)
+            elif (pos < old_pos) and (cur_pos < old_pos):
+                temp_name_to_dim[cur_name] = (cur_dt, cur_pos + 1)
+
+        # FIXME: renaming dimensions can recast PwAff as UnionPwAff. leaving
+        # this as-is until needed
+        if isinstance(new_isl_obj, isl.UnionPwAff):
+            raise NotImplementedError("UnionPwAff unsupported")
+
+        new_name_to_dim = temp_name_to_dim
+        new_name_to_dim[name] = (dim_type, pos)
+
+    return type(obj)(new_isl_obj, ordering)
+
+
+def _align_two(obj1: NamedIslTypeT,
+               obj2: NamedIslTypeT) -> Sequence[NamedIslTypeT]:
+    """
+    Aligns the spaces and name-to-dimension mappings of `obj1` and `obj2` so
+    that they are compatible for *named* set operations. `obj2` will first be
+    aligned to `obj1`, then `obj1` will be aligned to the result of the first
+    alignment.
+    """
+    if type(obj1) != type(obj2):
+        raise ValueError(
+            "Alignment requires both objects are of the same type. Got "
+           f"type(obj1) = {type(obj1)} and type(obj2) = {type(obj2)}"
+        )
+
+    ordering = _find_joint_name_to_dim(obj2._name_to_dim, obj1._name_to_dim)
+
+    obj2 = _align_space(obj2, ordering)
+    obj1 = _align_space(obj1, ordering)
+
+    return obj1, obj2
+
+
+def _align_and_apply_op(
+        obj1: NamedIslTypeT,
+        obj2: NamedIslTypeT,
+        op: Callable[[NamedIslTypeT, NamedIslTypeT], IslTypeT]
```

Comment:
These `NamedIslTypeT` are missing a type parameter that would have allowed you to precisely type `op`.